### PR TITLE
[#290] connecting existing code related to review feature

### PR DESCRIPTION
**Pull request by millevlada**
part 2 - UI for performing review and applying review on backend
- Separated UI for taking test and reviewing test - since two functionalities diverged too much in UI logic.
- Added awarded points per test question in model in order to have better control/overview over review awarded points.
- Implemented endpoint to submit review results.
- Changes in points awarding for multiple questions -> auto review is done at the moment when taken test is submit, and also awarding strategy is slightly changed to allow finer grained partial success points.

### CODE REVIEW

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 1]
```diff
@@ -238,6 +247,47 @@ private void updateQuestionWithAnswers(TestQuestion testQuestion,
...
+
+    int maxPoints = question.getPoints();
+    int numberOfCorrectAnswersInQuestion = question.getNumberOfCorrectAnswers();
+    int numberOfWrongAnswersInQuestion = question.getAnswers().size() - numberOfCorrectAnswersInQuestion;
+
+    Set<Long> correctAnswers = new HashSet<>(
+        question.getAnswers().stream()
+                .filter(Answer::isCorrect)
+                .map(Answer::getId)
+                .collect(Collectors.toList())
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Why not just use `Collectors.toSet()` instead?

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 2]
```diff
@@ -238,6 +247,47 @@ private void updateQuestionWithAnswers(TestQuestion testQuestion,
...
+    float pointsPerCorrectQuestion = ((float) maxPoints) / numberOfCorrectAnswersInQuestion;
+    float achievedPoints = pointsPerCorrectQuestion * numberOfGivenCorrectAnswers;
+
+    if (numberOfWrongAnswersInQuestion > 0 && numberOfGivenWrongAnswers > 0) {
+      float pointsPerWrongQuestion = ((float) maxPoints) / numberOfWrongAnswersInQuestion;
+      achievedPoints -= pointsPerWrongQuestion * numberOfGivenWrongAnswers;
+    }
+
+    int roundedAchievedPoints = Math.round(achievedPoints);
+    testQuestion.setAchievedPoints(Math.min(maxPoints, Math.max(0, roundedAchievedPoints)));
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: It was a business logic decision to either assign all points if only and all correct answers are selected, or no points, if there was an error.

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    if (status == HttpStatus.OK) {
+      return builder.body(new ReviewTestDto(test));
+    }
+    return builder.build();
+  }
+
+  private HttpStatus checkAccessStatusForReviewer(String token, Test test) {
+
+    if (!tokenStore.isValidToken(token)) {
+      return HttpStatus.BAD_REQUEST; // 400
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: Don't we want to return UNAUTHORIZED here, instead of BAD_REQUEST?

Comment (stanislavgrujic) [reply web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: Also, I would expect that there is an interceptor validating the token (or Spring doing it), instead of having to check it in each secured endpoint code.

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    Test test = testService.findOne(testId);
+    HttpStatus status = checkAccessStatusForReviewer(tokenKey, test);
+    ResponseEntity.BodyBuilder builder = ResponseEntity.status(status);
+    if (status == HttpStatus.OK) {
+      return builder.body(new ReviewTestDto(test));
+    }
+    return builder.build();
+  }
+
+  private HttpStatus checkAccessStatusForReviewer(String token, Test test) {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: Could you please move this method below the `submitReview()`, since it also invokes it, and it is easier to read the code if the called method is below the caller method.

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  public ResponseEntity<Void> submitReview(
+      @RequestBody SubmitReviewRequest submitReviewRequest,
+      @RequestHeader(required = false, name = AUTHENTICATION_TOKEN_KEY) String tokenKey) {
+    Map<Long, Integer> scores = submitReviewRequest.getScores();
+    Long testId = submitReviewRequest.getTestId();
+
+    Test test = testService.findOne(testId);
+    HttpStatus status = checkAccessStatusForReviewer(tokenKey, test);
+    ResponseEntity.BodyBuilder builder = ResponseEntity.status(status);
+    if (status == HttpStatus.OK) {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: It is strange that we first have the status of the request determined, and only after the actual logic. Could you please refactor, so that it follows the proper processing order?

File: web/src/main/java/eu/execom/testgenerator/web/dto/ReviewTestDto.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  public ReviewTestDto(Test test) {
+    super(test);
+    if (test != null) {
+
+      List<TestQuestion> testQuestions = test.getTestQuestions();
+      this.mapQuestionAnswer = new HashMap<>(testQuestions.size());
+
+      testQuestions.stream()
+        .map(ReviewTestQuestionAnswerDto::new)
+        .forEach(dto -> mapQuestionAnswer.put(dto.getQuestionId(), dto));
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/ReviewTestDto.java]: Why not use `Collectors.toMap()` instead?

File: web/src/test/java/eu/execom/testgenerator/web/controller/ReviewControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  @Mock
+  private TokenStore tokenStore;
+  @InjectMocks
+  private ReviewController reviewController;
+
+  private MockMvc mockMvc;
+
+  @Before
+  public void init() {
+    MockitoAnnotations.initMocks(this);
```
Comment (stanislavgrujic) [inline web/src/test/java/eu/execom/testgenerator/web/controller/ReviewControllerTest.java]: This is not needed, since annotation processing of `@InjectMocks` does the same thing.

File: web/src/main/frontend/src/components/ReviewTestQuestionComponent.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import ReviewCodeChallengeQuestion from './ReviewCodeChallengeQuestion';
+import Points from './Points';
+import questionTypes from '../base/constants/questionTypes';
+import {inject, observer} from 'mobx-react';
+
+@inject('completedTest')
+@observer
+class ReviewTestQuestionComponent extends React.Component {
+  componentDidMount() {
+    this.props.completedTest.addScore(
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewTestQuestionComponent.js]: Why is this here/needed?

File: web/src/main/frontend/src/components/ReviewTextAreaQuestion.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import React, {Component} from 'react';
+import {observer} from 'mobx-react';
+
+@observer
+class ReviewTextAreaQuestion extends Component {
+  render() {
+    return (
+      <React.Fragment>
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewTextAreaQuestion.js]: No need to wrap in `Fragment` if it's already in a `div`.

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import CodeChallengeQuestion from './CodeChallengeQuestion';
+import questionTypes from '../base/constants/questionTypes';
+import {inject, observer} from 'mobx-react';
+
+@inject('newTestForm')
+@observer
+class TakeTestQuestionComponent extends React.Component {
+  answers = this.props.newTestForm.$('answers').add();
+  render() {
+    this.answers.$('questionId').set(this.props.question.id);
```
Comment (iscekic) [inline web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: These two lines look like they belong in a lifecycle method.

File: web/src/main/frontend/src/components/ui/Checkbox.js [hunk 11]
```diff
@@ -2,13 +2,22 @@ import React from 'react';
 
 const Checkbox = props => {
   const answerTextClassName = `label_text ${
-    props.checked && !props.isCorrect ? 'wrong' : ''
+    (props.checked && !props.isCorrect) || (!props.checked && props.isCorrect)
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ui/Checkbox.js]: Use a variable.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java (hunk 1 above)
Comment (millevlada) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: tnx, I'll fix it on other places as well.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java (hunk 2 above)
Comment (millevlada) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Hm, ok, I was not aware of that. However, I think that other test platforms are doing partial points approach? I just added in meanwhile number of unit tests to document the awarding strategy... Maybe we could keep it as is now ;) ?

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java (hunk 3 above)
Comment (millevlada) [reply web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: I saw it was done like this on other existing places in code (i.e. TestController). I can change it to UNAUTHORIZED, it does seem more correct indeed.

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java (hunk 4 above)
Comment (millevlada) [reply web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: ok, it was originally used by other public method before submitReview. I'll move it lower.

File: web/src/main/java/eu/execom/testgenerator/web/dto/ReviewTestDto.java (hunk 6 above)
Comment (millevlada) [reply web/src/main/java/eu/execom/testgenerator/web/dto/ReviewTestDto.java]: cool, I was not aware of that. I'll fix it.

File: web/src/test/java/eu/execom/testgenerator/web/controller/ReviewControllerTest.java (hunk 7 above)
Comment (millevlada) [reply web/src/test/java/eu/execom/testgenerator/web/controller/ReviewControllerTest.java]: Saw it somewhere else... learning by copy/paste :) I'll fix it on other places too.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java (hunk 2 above)
Comment (stanislavgrujic) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Depends on the platform. OK, leave it, but please check next time before changing something in business logic flows. ;)

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java (hunk 3 above)
Comment (millevlada) [reply web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: About interceptor -> well, I do not know about that. This was also in other existing code. I'll try to research about such feature, but preferably in some separate PR, since this one unblock quite a lot of missing functionality.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java (hunk 2 above)
Comment (millevlada) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Ok... This time I was connecting some loose ends of both UI and backend, so I was actually not aware that I'm changing something decided until I stumbled into parts of code that were there but never connected with anything... Then I realized that what I coded was actually already existing, in some parts, done with different logic. But ok, we might make this configurable -> 'partial scoring' or 'all or nothing'. ;)

File: web/src/test/java/eu/execom/testgenerator/web/controller/ReviewControllerTest.java (hunk 7 above)
Comment (millevlada) [reply web/src/test/java/eu/execom/testgenerator/web/controller/ReviewControllerTest.java]: Ok, fixed in 5-6 other classes. Actually, there it was missing @RunWith(MockitoJUnitRunner.class), so probably therefore it was causing confusion.

File: web/src/main/frontend/src/base/rest/reviewApi.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import {get, put, withLoader} from './requests';
+
+class reviewApi {
+  getTest(testId) {
+    return withLoader(get(`/reviews/${testId}`));
+  }
+
+  submitReview(review) {
+    return withLoader(put('/reviews/submit', review));
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/base/rest/reviewApi.js]: Well, by doing 'submitReview' we actually do update of already existing test. I can rename on UI part to 'applyReview' if that is better?

File: web/src/main/frontend/src/base/stores/CompletedTestStore.js [hunk 13]
```diff
@@ -34,6 +35,16 @@ class CompletedTestStore {
   get testQuestions() {
     return this.test.questions;
   }
+
+  getGivenAnswerForQuestion(questionId) {
+    return this.test.mapQuestionAnswer[questionId];
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/base/stores/CompletedTestStore.js]: For me as 'code user' seems natural to ask store for given answer details for specified question id, and to hide entire map from user code? 
I can rename this map to answerForQuestion, that seems better indeed.

File: web/src/main/frontend/src/components/Points.js [hunk 14]
```diff
@@ -10,35 +10,39 @@ class Points extends React.Component {
...
-        : `0/${this.props.points}`;
+    if (fixed || points !== null) {
+      inputValue = `${this.props.points} / ${this.props.maxpoints}`;
     }
 
     this.setState({inputValue});
   };
 
-  createInputFieldProps(fixed, correct, valuated) {
+  createInputFieldProps(fixed, points, maxpoints) {
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/Points.js]: 'fixed' means that points are fixed -> user can not manually add/change them. That is optional attribute (just present or not present), it is not really a boolean.

[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/Points.js]: I mean:
<Points fixed otherAttributes... />
or 
<Points otherAttributes...>

it is not like:
<Points fixed={true} otherAttributes... />
or
<Points fixed={false} otherAttributes... />

File: web/src/main/frontend/src/components/Points.js [hunk 15]
```diff
@@ -10,35 +10,39 @@ class Points extends React.Component {
...
+            : 'correct'
+        : ''
     }`;
-    inputProps.readOnly = fixed || null;
-    inputProps.onChange = valuated && this.onChange;
-    inputProps.onBlur = valuated && this.onBlur;
-    inputProps.onFocus = valuated && this.onFocus;
-    inputProps.onKeyDown = valuated && this.onKeyDown;
-    inputProps.placeholder = `0/${this.props.points}`;
+    inputProps.readOnly = !!fixed;
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/Points.js]: That operator (!!) I got as check style warning solution hint on commit. ;) It is not really 'boolean'.

File: web/src/main/frontend/src/components/Points.js [hunk 16]
```diff
@@ -76,15 +80,17 @@ class Points extends React.Component {
...
     if (event.target.value) {
-      this.setState({inputValue: event.target.value + `/${this.props.points}`});
+      this.setState({
+        inputValue: event.target.value + ` / ${this.props.maxpoints}`
+      });
     }
   };
 
   render() {
     const inputFieldProps = this.createInputFieldProps(
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/Points.js]: no idea... this is just old existing code? ;) If really needed, I can move it - I need to read first what getDerivedStateFromProps is... ;)

File: web/src/main/frontend/src/components/ReviewTest.js [hunk 17]
```diff
@@ -2,14 +2,31 @@ import React from 'react';
...
+      this.props.dashboard.fetchDashboardData();
+    });
+  };
+
   render() {
     const questionsToReview = this.props.completedTest.testQuestions.map(
-      (question, index) => getComponentByQuestionType(question, index, true)
+      (question, index) => (
+        <ReviewTestQuestionComponent
+          key={index}
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/ReviewTest.js]: I added that just to fixing ugly large warning given in console:
index.js:2178 Warning: Each child in an array or iterator should have a unique "key" prop.

Check the render method of `Test`. It was passed a child from ReviewTest. See https://fb.me/react-warning-keys for more information.
...

So, maybe be should just keep that key as is?

File: web/src/main/frontend/src/components/ReviewTestQuestionComponent.js (hunk 8 above)
Comment (millevlada) [reply web/src/main/frontend/src/components/ReviewTestQuestionComponent.js]: In order to update overall score given in the result input under the test... And also it 'registers' existing values of achieved points in map that is posted to backend later, even when user does not change any values.

File: web/src/main/frontend/src/components/ReviewTextAreaQuestion.js (hunk 9 above)
Comment (millevlada) [reply web/src/main/frontend/src/components/ReviewTextAreaQuestion.js]: Ok, this was copy of existing component (TextAreaQuestion), I'll fix it there too.

File: web/src/main/frontend/src/components/ui/Checkbox.js (hunk 11 above)
Comment (millevlada) [reply web/src/main/frontend/src/components/ui/Checkbox.js]: fixed

File: web/src/main/frontend/src/components/ReviewTextAreaQuestion.js (hunk 9 above)
Comment (millevlada) [reply web/src/main/frontend/src/components/ReviewTextAreaQuestion.js]: fixed

File: web/src/main/frontend/src/components/Points.js [hunk 18]
```diff
@@ -76,15 +80,17 @@ class Points extends React.Component {
 
   onBlur = event => {
     if (event.target.value) {
-      this.setState({inputValue: event.target.value + `/${this.props.points}`});
+      this.setState({
+        inputValue: event.target.value + ` / ${this.props.maxpoints}`
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/Points.js]: fixed

File: web/src/main/frontend/src/components/Points.js [hunk 19]
```diff
@@ -10,35 +10,39 @@ class Points extends React.Component {
   inputRef = React.createRef();
 
   componentDidMount() {
-    if (this.props.isCorrect) {
-      this.props.completedTest.addScore(this.props.points, this.props.id);
-    }
-
-    this.getValueForInput(this.props.fixedScore, this.props.isCorrect);
+    this.getValueForInput(
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/Points.js]: fixed

File: web/src/main/frontend/src/base/stores/CompletedTestStore.js [hunk 20]
```diff
@@ -34,6 +35,16 @@ class CompletedTestStore {
...
+  }
+
+  @action.bound
+  submitReview() {
+    const review = {scores: this.scores, testId: this.test.id};
```
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/base/stores/CompletedTestStore.js]: fixed

File: web/src/main/frontend/src/base/stores/CompletedTestStore.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/base/stores/CompletedTestStore.js]: fixed name

File: web/src/main/frontend/src/base/rest/reviewApi.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (iscekic) [reply web/src/main/frontend/src/base/rest/reviewApi.js]: `updateReview` sounds ok to me 😄

File: web/src/main/frontend/src/components/ReviewTest.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (iscekic) [reply web/src/main/frontend/src/components/ReviewTest.js]: If question has an `id` of some sort, we should use that as `key`. If not, leave it as is.

File: web/src/main/frontend/src/components/Points.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (iscekic) [reply web/src/main/frontend/src/components/Points.js]: Leave it for now, we can create an improvement task. 😄

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js (hunk 10 above)
Comment (iscekic) [reply web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: Ok, leave a `TODO`

File: web/src/main/frontend/src/components/Points.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (iscekic) [reply web/src/main/frontend/src/components/Points.js]: In your first two examples:
1) `fixed === true`
2) `fixed === undefined` - `undefined` is falsy

So while `fixed` isn't really boolean, we can treat it like it is. 😄

File: web/src/main/frontend/src/components/ReviewCodeChallengeQuestion.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+      ...this.state,
+      syntax: select
+    });
+  };
+
+  updateCode = (editor, data, value) => this.setState({value});
+
+  render() {
+    return (
+      <React.Fragment>
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewCodeChallengeQuestion.js]: No need to wrap.

File: web/src/main/frontend/src/components/ReviewMultipleChoiceQuestion.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+          option.id
+        )}
+        label={option.description}
+      />
+    );
+  });
+
+  render() {
+    return (
+      <React.Fragment>
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewMultipleChoiceQuestion.js]: No need to wrap.

File: web/src/main/frontend/src/components/ReviewTest.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/ReviewTest.js]: Ok, it has id, I'll change it then to question.id

File: web/src/main/frontend/src/components/TakeTestQuestionComponent.js (hunk 10 above)
Comment (millevlada) [reply web/src/main/frontend/src/components/TakeTestQuestionComponent.js]: deal

File: web/src/main/frontend/src/components/Points.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (millevlada) [reply web/src/main/frontend/src/components/Points.js]: Ok, fixed, it works with just boolean compare -> thanks. ;)

File: web/src/main/frontend/src/components/ReviewCodeChallengeQuestion.js (hunk 21 above)
Comment (millevlada) [reply web/src/main/frontend/src/components/ReviewCodeChallengeQuestion.js]: ok, fixed in old component too

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java (hunk 3 above)
Comment (millevlada) [reply web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: Fixed ReviewController, it now uses 'regular' authentication, and logged in user we now read from Principal.

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java [hunk 23]
```diff
@@ -30,54 +29,60 @@
...
+      Authentication authentication = (Authentication) principal;
+      if (authentication.getPrincipal() instanceof User) {
+        return (User) authentication.getPrincipal();
+      }
     }
 
-    return HttpStatus.OK;
+    log.error("Unsupported principal, unable to extract logged in user! " + principal);
+
+    return null;
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: Better throw `AccessDeniedException` which will result in returning the same UNAUTHORIZED status code, but you will avoid having the null checks.

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java [hunk 24]
```diff
@@ -30,54 +29,60 @@
...
+  private User resolveLoggedInUser(Principal principal) {
+    if (principal instanceof Authentication) {
+      Authentication authentication = (Authentication) principal;
+      if (authentication.getPrincipal() instanceof User) {
+        return (User) authentication.getPrincipal();
+      }
     }
 
-    return HttpStatus.OK;
+    log.error("Unsupported principal, unable to extract logged in user! " + principal);
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: It is more efficient to use parameter placeholder than string concatenation with logger -
```
log.error("Unsupported principal, unable to extract logged in user! {} ", principal);
```

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java (hunk 23 above)
Comment (millevlada) [reply web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: great!

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java (hunk 24 above)
Comment (millevlada) [reply web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: ok, I'll fix it
